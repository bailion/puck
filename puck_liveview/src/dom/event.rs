#[derive(Clone)]
pub struct ClickEvent;

#[derive(Clone)]
pub struct SubmitEvent;

#[derive(Clone)]
pub struct InputEvent {
    pub value: String,
}
