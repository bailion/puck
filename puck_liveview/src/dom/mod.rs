//! The server-side virtual DOM.

pub mod element;
pub mod event;
pub mod listener;
