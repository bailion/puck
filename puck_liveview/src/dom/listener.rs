use std::fmt::Debug;

use super::event::{ClickEvent, InputEvent, SubmitEvent};

/// A DOM listener.

pub enum Listener<INPUT> {
    Click {
        call: Box<dyn Fn(ClickEvent) -> INPUT>,
    },
    Submit {
        call: Box<dyn Fn(SubmitEvent) -> INPUT>,
    },
    Input {
        call: Box<dyn Fn(InputEvent) -> INPUT>,
    },
}

impl<INPUT> Debug for Listener<INPUT> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Listener").finish()
    }
}

impl<INPUT> From<Box<dyn Fn(ClickEvent) -> INPUT>> for Listener<INPUT> {
    fn from(call: Box<dyn Fn(ClickEvent) -> INPUT>) -> Self {
        Self::Click { call }
    }
}

impl<INPUT> From<Box<dyn Fn(SubmitEvent) -> INPUT>> for Listener<INPUT> {
    fn from(call: Box<dyn Fn(SubmitEvent) -> INPUT>) -> Self {
        Self::Submit { call }
    }
}

impl<INPUT> From<Box<dyn Fn(InputEvent) -> INPUT>> for Listener<INPUT> {
    fn from(call: Box<dyn Fn(InputEvent) -> INPUT>) -> Self {
        Self::Input { call }
    }
}

impl<INPUT> Listener<INPUT> {
    #[inline]
    pub fn js_event(&self) -> &'static str {
        match self {
            Listener::Click { call: _ } => "click",
            Listener::Submit { call: _ } => "submit",
            Listener::Input { call: _ } => "input",
        }
    }
}

#[derive(Debug, Clone)]
pub struct ListenerRef {
    listener_name: String,
    js_event: String,
}

impl ListenerRef {
    pub fn new(listener_name: impl ToString, js_event: impl ToString) -> Self {
        Self {
            listener_name: listener_name.to_string(),
            js_event: js_event.to_string(),
        }
    }
    /// Get a reference to the listener ref's listener name.
    pub fn listener_name(&self) -> &str {
        self.listener_name.as_str()
    }

    /// Get a reference to the listener ref's js event.
    pub fn js_event(&self) -> &str {
        self.js_event.as_str()
    }
}
