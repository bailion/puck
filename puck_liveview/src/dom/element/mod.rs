//! Create HTML elements. Unfortunately at the moment this is the API – in the future we'll provide
//! something a bit nicer to work with.

use std::{borrow::Cow, collections::HashMap};

use super::listener::ListenerRef;

pub mod diff;
pub mod orchestrator;
pub mod render;

/// An HTML element.
#[derive(Builder, Clone, Default, Debug, Eq, PartialEq)]
pub struct Element {
    /// The id includes the ID of this element and all the parent elements.
    ///
    /// This might need to be changed in the future.
    pub id: Vec<u32>,
    #[builder(setter(into))]
    pub name: Cow<'static, str>,
    pub attributes: HashMap<Cow<'static, str>, Cow<'static, str>>,
    pub listeners: Vec<ListenerRef>,
    pub children: Vec<Element>,
    pub text: Option<Cow<'static, str>>,
    pub key: Option<String>,
}

impl Element {
    /// Return the unique ID of this element.
    ///
    /// At the moment this allocates way too much – at some point we should update this structure
    /// to use more indirection.
    fn id(&self) -> String {
        self.id
            .iter()
            .map(ToString::to_string)
            .map(|x| x + "-")
            .collect::<String>()
    }

    /// This function returns a builder type for this struct.
    pub fn build() -> ElementBuilder {
        ElementBuilder::default()
    }

    /// Create an `Element` using `Default::default()`, except for the API specified.
    pub fn default_with_id(id: Vec<u32>) -> Self {
        Self {
            id,
            ..Default::default()
        }
    }
}
