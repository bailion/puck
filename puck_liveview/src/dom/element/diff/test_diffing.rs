use maplit::hashmap;

use crate::dom::{element::Element, listener::ListenerRef};

#[test]
fn test_name_change() {
    let old = Element {
        id: vec![0],
        name: "div".into(),
        ..Default::default()
    };

    let new = Element {
        id: vec![0],
        name: "p".into(),
        ..Default::default()
    };

    let cs = old.diff(Some(&new));

    insta::assert_debug_snapshot!("test_name_change", cs);
}

#[test]
fn test_single_element_attribute_change() {
    let old = Element {
        id: vec![0],
        name: "div".into(),
        attributes: hashmap! {
            "class".into() => "one".into(),
            "attribute-which-doesn-t-exist-after-diffing".into() => "1".into()
        },
        ..Default::default()
    };

    let new = Element {
        id: vec![0],
        name: "div".into(),
        attributes: hashmap! {
            "class".into() => "two".into(),
            "new-attribute-added-after-diffing".into() => "value".into()
        },
        ..Default::default()
    };

    let cs = old.diff(Some(&new));

    insta::assert_debug_snapshot!("test_single_element_attribute_change", cs);
}

#[test]
fn test_single_element_text_change() {
    let old = Element {
        id: vec![0],
        name: "p".into(),
        text: Some("the cat sat on the mat".into()),
        ..Default::default()
    };

    let new = Element {
        id: vec![0],
        name: "p".into(),
        text: Some("the mat sat on the cat".into()),
        ..Default::default()
    };

    let c = old.diff(Some(&new));
    insta::assert_debug_snapshot!("test_single_element_text_change", c);
}

#[test]
fn test_add_child_change() {
    let old = Element {
        id: vec![0],
        name: "div".into(),
        children: vec![Element {
            id: vec![0, 0],
            key: Some("a".to_string()),
            name: "p".into(),
            text: Some("the cat sat on the mat".into()),
            ..Default::default()
        }],
        ..Default::default()
    };

    let new = Element {
        id: vec![0],
        name: "div".into(),
        children: vec![
            Element {
                id: vec![0, 0],
                name: "p".into(),
                key: Some("a".to_string()),
                text: Some("the cat sat on the mat".into()),
                ..Default::default()
            },
            Element {
                id: vec![0, 1],
                name: "p".into(),
                key: Some("b".to_string()),
                text: Some("the mat sat on the cat".into()),
                ..Default::default()
            },
        ],
        ..Default::default()
    };

    let c = old.diff(Some(&new));
    insta::assert_debug_snapshot!("test_add_child_change", c);
}

#[test]
fn test_add_child_before_change() {
    let old = Element {
        id: vec![0],
        name: "div".into(),
        children: vec![Element {
            id: vec![0, 0],
            key: Some("a".to_string()),
            name: "p".into(),
            text: Some("the cat sat on the mat".into()),
            ..Default::default()
        }],
        ..Default::default()
    };

    let new = Element {
        id: vec![0],
        name: "div".into(),
        children: vec![
            Element {
                id: vec![0, 0],
                name: "p".into(),
                key: Some("b".to_string()),
                text: Some("the mat sat on the cat".into()),
                ..Default::default()
            },
            Element {
                id: vec![0, 1],
                name: "p".into(),
                key: Some("a".to_string()),
                text: Some("the cat sat on the mat".into()),
                ..Default::default()
            },
        ],
        ..Default::default()
    };

    let c = old.diff(Some(&new));
    insta::assert_debug_snapshot!("test_add_child_before_change", c);
}

#[test]
fn test_more_complex_diff() {
    let old = Element {
        id: vec![0],
        name: std::borrow::Cow::Borrowed("div"),
        attributes: hashmap! {
            "class".into() => "message-list".into(),
        },
        listeners: vec![],
        children: vec![
            Element {
                id: vec![0, 0],
                name: std::borrow::Cow::Borrowed("div"),
                attributes: hashmap! {},
                listeners: vec![],
                children: vec![Element {
                    id: vec![0, 0, 0],
                    name: std::borrow::Cow::Borrowed("input"),
                    attributes: hashmap! {},
                    listeners: vec![ListenerRef::new("msg-input", "input")],
                    children: vec![],
                    text: None,
                    key: None,
                }],
                text: None,
                key: None,
            },
            Element {
                id: vec![0, 1],
                name: std::borrow::Cow::Borrowed("div"),
                attributes: hashmap! {},
                listeners: vec![],
                children: vec![Element {
                    id: vec![0, 1, 0],
                    name: std::borrow::Cow::Borrowed("button"),
                    attributes: hashmap! {},
                    listeners: vec![ListenerRef::new("msg-submit", "click")],
                    children: vec![],
                    text: Some(std::borrow::Cow::Borrowed("Send message")),
                    key: None,
                }],
                text: None,
                key: None,
            },
        ],
        text: None,
        key: None,
    };
    let new = Element {
        id: vec![0],
        name: std::borrow::Cow::Borrowed("div"),
        attributes: hashmap! {
            "class".into() => "message-list".into(),
        },
        listeners: vec![],
        children: vec![
            Element {
                id: vec![0, 0],
                name: std::borrow::Cow::Borrowed("div"),
                attributes: hashmap! {},
                listeners: vec![],
                children: vec![Element {
                    id: vec![0, 0, 0],
                    name: std::borrow::Cow::Borrowed("input"),
                    attributes: hashmap! {},
                    listeners: vec![ListenerRef::new(
                        "msg-input".to_string(),
                        "input".to_string(),
                    )],
                    children: vec![],
                    text: None,
                    key: None,
                }],
                text: None,
                key: None,
            },
            Element {
                id: vec![0, 1],
                name: std::borrow::Cow::Borrowed("div"),
                attributes: hashmap! {},
                listeners: vec![],
                children: vec![Element {
                    id: vec![0, 1, 0],
                    name: std::borrow::Cow::Borrowed("button"),
                    attributes: hashmap! {},
                    listeners: vec![ListenerRef::new(
                        "msg-submit".to_string(),
                        "click".to_string(),
                    )],
                    children: vec![],
                    text: Some(std::borrow::Cow::Borrowed("Send message")),
                    key: None,
                }],
                text: None,
                key: None,
            },
            Element {
                id: vec![0, 2],
                name: std::borrow::Cow::Borrowed("div"),
                attributes: hashmap! {
                    "class".into() =>  "message-container".into(),
                },
                listeners: vec![],
                children: vec![
                    Element {
                        id: vec![0, 2, 0],
                        name: std::borrow::Cow::Borrowed("p"),
                        attributes: hashmap! {
                            "class".into() => "message-sent-at".into(),
                        },
                        listeners: vec![],
                        children: vec![],
                        text: Some(std::borrow::Cow::Borrowed("1970-01-25 06:34:13")),
                        key: None,
                    },
                    Element {
                        id: vec![0, 2, 1],
                        name: std::borrow::Cow::Borrowed("p"),
                        attributes: hashmap! {
                            "class".into() => "message-author".into(),
                        },
                        listeners: vec![],
                        children: vec![],
                        text: Some(std::borrow::Cow::Borrowed("[username not set]")),
                        key: None,
                    },
                    Element {
                        id: vec![0, 2, 2],
                        name: std::borrow::Cow::Borrowed("p"),
                        attributes: hashmap! {
                            "class".into() => "message-contents".into(),
                        },
                        listeners: vec![],
                        children: vec![],
                        text: Some(std::borrow::Cow::Borrowed("sending message")),
                        key: None,
                    },
                ],
                text: None,
                key: None,
            },
        ],
        text: None,
        key: None,
    };

    let c = old.diff(Some(&new));
    insta::assert_debug_snapshot!(c);
}

#[test]
fn test_delete_child_change() {}
