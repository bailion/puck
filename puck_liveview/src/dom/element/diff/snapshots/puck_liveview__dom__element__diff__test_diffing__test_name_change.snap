---
source: puck_liveview/src/dom/element/diff/test_diffing.rs
expression: cs

---
Changeset {
    ops: [
        Op {
            id: "0-",
            instruction: SetTagName {
                name: "p",
            },
        },
    ],
}
